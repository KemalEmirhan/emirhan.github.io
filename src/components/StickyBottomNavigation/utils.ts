import { NAVIGATIONS } from '@/constants/navigations';

export const getActiveLinkIndex = (pathname: string) =>
  NAVIGATIONS.findIndex(navigation => navigation.href === pathname);

export const calculateDotPosition = (
  navElement: HTMLElement,
  activeIndex: number,
  direction: 'horizontal' | 'vertical' = 'horizontal'
): number => {
  const links = navElement.querySelectorAll('a');
  const activeLink = links[activeIndex];

  if (!activeLink) return 0;

  const navRect = navElement.getBoundingClientRect();
  const linkRect = activeLink.getBoundingClientRect();

  return direction === 'horizontal'
    ? linkRect.left - navRect.left + linkRect.width / 2
    : linkRect.top - navRect.top + linkRect.height / 2;
};
