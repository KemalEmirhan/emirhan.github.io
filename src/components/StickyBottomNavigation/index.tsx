'use client';

import Link from 'next/link';
import { usePathname } from 'next/navigation';
import React, { useState, useCallback } from 'react';
import { motion } from 'motion/react';
import { getActiveLinkIndex, calculateDotPosition } from './utils';
import { NAVIGATIONS } from '@/constants/navigations';
import useTracking from '@/hooks/useTracking';

const StickyBottomNavigation = () => {
  const pathname = usePathname();
  const [dotX, setDotX] = useState(20); // Initially set to the center of the first link
  const { clickEvent } = useTracking('Sticky Bottom Navigation');
  const activeIndex = getActiveLinkIndex(pathname);

  const navRef = useCallback(
    (navElement: HTMLElement | null) => {
      if (!navElement) return;

      const centerX = calculateDotPosition(navElement, activeIndex);
      setDotX(centerX);
    },
    [activeIndex]
  );

  const handleClick = (href: string) => {
    clickEvent({
      category: 'navigation',
      label: href,
      value: activeIndex,
    });
  };

  return (
    <div className='fixed lg:hidden bottom-8 left-1/2 -translate-x-1/2 bg-white rounded-3xl shadow-lg px-4 py-2 drop-shadow-lg border border-gray-500 w-fit z-50'>
      <nav
        ref={navRef}
        className='flex justify-evenly items-center gap-4 relative'
      >
        <motion.div
          className='absolute w-2 h-2 bg-blue-500 rounded-full bottom-[-4px]'
          initial={false}
          animate={{ left: dotX - 4 }}
          transition={{ type: 'spring', stiffness: 300, damping: 25 }}
        />

        {NAVIGATIONS.map(item => (
          <Link
            key={item.href}
            href={item.href}
            className='flex flex-col items-center'
            onClick={() => handleClick(item.href)}
          >
            <span
              className='text-gray-500 data-[active=true]:text-gray-900 transition-colors duration-150'
              data-active={pathname === item.href}
            >
              {item.label}
            </span>
          </Link>
        ))}
      </nav>
    </div>
  );
};

export default StickyBottomNavigation;
