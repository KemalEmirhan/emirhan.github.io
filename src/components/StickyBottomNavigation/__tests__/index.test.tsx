import { afterAll, describe, expect, test, vi } from 'vitest';
import { render } from '@testing-library/react';
import StickyBottomNavigation from '../';

const mockPathname = vi.fn();

vi.mock('next/navigation', () => ({
  usePathname: () => mockPathname(),
}));

describe('StickyBottomNavigation', () => {
  afterAll(() => {
    vi.clearAllMocks();
  });

  test('should render', () => {
    mockPathname.mockReturnValue('/');
    const { container } = render(<StickyBottomNavigation />);
    expect(container).toMatchSnapshot();
  });

  test('should handle /', () => {
    mockPathname.mockReturnValue('/');
    const { container, getByText } = render(<StickyBottomNavigation />);
    expect(getByText('About')).toHaveAttribute('data-active', 'true');
    expect(container).toMatchSnapshot();
  });

  test('should handle /socials', () => {
    mockPathname.mockReturnValue('/socials');
    const { container, getByText } = render(<StickyBottomNavigation />);
    expect(getByText('Socials')).toHaveAttribute('data-active', 'true');
    expect(container).toMatchSnapshot();
  });
});
