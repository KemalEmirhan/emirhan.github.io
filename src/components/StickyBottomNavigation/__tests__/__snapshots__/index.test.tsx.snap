// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StickyBottomNavigation > should handle / 1`] = `
<div>
  <div
    class="fixed lg:hidden bottom-8 left-1/2 -translate-x-1/2 bg-white rounded-3xl shadow-lg px-4 py-2 drop-shadow-lg border border-gray-500 w-fit z-50"
  >
    <nav
      class="flex justify-evenly items-center gap-4 relative"
    >
      <div
        class="absolute w-2 h-2 bg-blue-500 rounded-full bottom-[-4px]"
        style="left: 16px;"
      />
      <a
        class="flex flex-col items-center"
        href="/"
      >
        <span
          class="text-gray-500 data-[active=true]:text-gray-900 transition-colors duration-150"
          data-active="true"
        >
          About
        </span>
      </a>
      <a
        class="flex flex-col items-center"
        href="/socials"
      >
        <span
          class="text-gray-500 data-[active=true]:text-gray-900 transition-colors duration-150"
          data-active="false"
        >
          Socials
        </span>
      </a>
    </nav>
  </div>
</div>
`;

exports[`StickyBottomNavigation > should handle /socials 1`] = `
<div>
  <div
    class="fixed lg:hidden bottom-8 left-1/2 -translate-x-1/2 bg-white rounded-3xl shadow-lg px-4 py-2 drop-shadow-lg border border-gray-500 w-fit z-50"
  >
    <nav
      class="flex justify-evenly items-center gap-4 relative"
    >
      <div
        class="absolute w-2 h-2 bg-blue-500 rounded-full bottom-[-4px]"
        style="left: 16px;"
      />
      <a
        class="flex flex-col items-center"
        href="/"
      >
        <span
          class="text-gray-500 data-[active=true]:text-gray-900 transition-colors duration-150"
          data-active="false"
        >
          About
        </span>
      </a>
      <a
        class="flex flex-col items-center"
        href="/socials"
      >
        <span
          class="text-gray-500 data-[active=true]:text-gray-900 transition-colors duration-150"
          data-active="true"
        >
          Socials
        </span>
      </a>
    </nav>
  </div>
</div>
`;

exports[`StickyBottomNavigation > should render 1`] = `
<div>
  <div
    class="fixed lg:hidden bottom-8 left-1/2 -translate-x-1/2 bg-white rounded-3xl shadow-lg px-4 py-2 drop-shadow-lg border border-gray-500 w-fit z-50"
  >
    <nav
      class="flex justify-evenly items-center gap-4 relative"
    >
      <div
        class="absolute w-2 h-2 bg-blue-500 rounded-full bottom-[-4px]"
        style="left: 16px;"
      />
      <a
        class="flex flex-col items-center"
        href="/"
      >
        <span
          class="text-gray-500 data-[active=true]:text-gray-900 transition-colors duration-150"
          data-active="true"
        >
          About
        </span>
      </a>
      <a
        class="flex flex-col items-center"
        href="/socials"
      >
        <span
          class="text-gray-500 data-[active=true]:text-gray-900 transition-colors duration-150"
          data-active="false"
        >
          Socials
        </span>
      </a>
    </nav>
  </div>
</div>
`;
